google-cloud-spanner/src/main/java/com/google/cloud/spanner/Database.java

```diff
@@ -88,6 +88,20 @@ public OperationFuture<Void, UpdateDatabaseDdlMetadata> updateDdl(
     return dbClient.updateDatabaseDdl(instance(), database(), statements, operationId);
   }
 
+  /**
+   * Updates a database in a Cloud Spanner instance
+   *
+   * @param database Database object to set configuration options such as proto_descriptors.
+   * @param statements DDL statements to run while updating the database.
+   * @param operationId Operation id assigned to this operation. If null, system will autogenerate
+   *     one. This must be unique within a database abd must be a valid identifier
+   *     [a-zA-Z][a-zA-Z0-9_]*.
+   */
+  public OperationFuture<Void, UpdateDatabaseDdlMetadata> updateDdl(
```

Comment:
Sorry, I didn't notice this during the last review, but this seems a little strange. It is a method in the `Database` class and takes a `Database` instance as an argument. Why does it not just operate on this database?

Reply:
The reason it cannot operate on this database is because a user cannot set `ProtoDescriptor` directly to Database Class as `setProtoDescriptors` method belongs to `Database.Builder`. Below is a small example to illustrate this,
```
Database database = client.getDatabase(INSTANCE_ID, DB_ID);

// If a user wants to update the ProtoDescriptor field then, 
// the below line cannot be written as setProtoDescriptors method does not belong to Database class.
database.setProtoDescriptors("descriptors.pb"); 

// This does not update proto descriptor in the database because the field was not updated.
database.updateDdl(Collections.singletonList("DROP TABLE Singer"), null); 
```
So a new Database instance has to be passed to update the ProtoDescriptor field. However, I agree that this is incorrect and confusing at same time.

**Another Approach**
@olavloite  Please validate if below is more meaningful to do.

Expose a new method `toBuilder()` in `Database.java` that return a Builder of that `Database` instance. This allows users to `setProtoDescriptor` by calling the builder.
The change in `Database.java` is
```
public class Database extends DatabaseInfo {
  public Builder toBuilder() {
      return new Builder(this);
    }


  public OperationFuture<Void, UpdateDatabaseDdlMetadata> updateDdl(
      Iterable<String> statements, String operationId) throws SpannerException {
     // The updateDatabaseDdl is called with 'this' as argument instead of instance(), database()
    return dbClient.updateDatabaseDdl(this, statements, operationId);
  }
}
```
The user can use it like below
```
Database database = client.getDatabase(INSTANCE_ID, DB_ID);
// The below database instance has the updated ProtoDescriptor field
database = database.toBuilder().setProtoDescriptors("descriptors.pb").build();
// Update works fine
database.updateDdl(Collections.singletonList("DROP TABLE Singer"), null);
```

Reply:
Hmmm.... If the `protoDescriptors` field is not something that is actually part of the `Database` class and thereby also not stored with the database, and is only needed for the update method, then we need to add it as an argument to the method afte rall. Other approaches would be non-idiomatic Java.

So the signature of the method would then become `Database#updateDatabaseDdl(Iterable<String> statements, String protoDescriptors, String operationId)`. Another option would be to create a `DatabaseUpdateOptions` class that contains all possible update options, the first one being `protoDescriptors` (that would make it more easily reusable for additional options in the future).

Reply:
I agree with you @olavloite. Creating a new `DatabaseUpdateOptions` class that can update options will help us reuse in future. I will take this effort as part of the next PR. I will tag this comment in the other PR later to understand the need of that change.